lib/server.js

```diff
@@ -152,39 +152,32 @@ class Server extends KarmaEventEmitter {
         this._injector.invoke(watcher.watch)
       }
 
-      const startWebServer = () => {
-        webServer.listen(config.port, config.listenAddress, () => {
-          this.log.info(`Karma v${constant.VERSION} server started at ${config.protocol}//${config.listenAddress}:${config.port}${config.urlRoot}`)
-
-          this.emit('listening', config.port)
-          if (config.browsers && config.browsers.length) {
-            this._injector.invoke(launcher.launch, launcher).forEach((browserLauncher) => {
-              singleRunDoneBrowsers[browserLauncher.id] = false
-            })
-          }
-          if (this.loadErrors.length > 0) {
-            this.log.error('Found %d load error%s', this.loadErrors.length, this.loadErrors.length === 1 ? '' : 's')
-            process.exitCode = 1
-            process.kill(process.pid, 'SIGINT')
-          }
-        })
-      }
-
-      if (fs.existsSync(karmaJsPath) && fs.existsSync(contextJsPath)) {
-        startWebServer()
-      } else {
-        this.log.info('Front-end scripts not present. Compiling...')
-        Promise.all([
-          bundleResource(path.join(__dirname, '/../client/main.js'), karmaJsPath),
-          bundleResource(path.join(__dirname, '/../context/main.js'), contextJsPath)
-        ])
-          .then(startWebServer)
-          .catch((error) => {
-            this.log.error('Front-end script compile failed with error: ' + error)
-            process.exitCode = 1
-            process.kill(process.pid, 'SIGINT')
+      return Promise.all([
+        bundleResource(path.join(__dirname, '/../client/main.js'), path.join(__dirname, '/../static/karma.js')),
```

Comment:
I think the code would be clearer if you create  function to compute the paths:
```js
  bundleResource(['/../client/main.js', '/../static/karma.js'].map(absPath));
```
If you also pass the `fs`:
```js
  bundleResource(fs, ['/../client/main.js', '/../static/karma.js'].map(absPath));
```
then we can unit test the `bundleResource()` properly.

Reply:
Maybe a bit other way, but done :)